python/plugins/processing/algs/qgis/RasterSampling.py

```diff
@@ -0,0 +1,163 @@
+
+from processing.algs.qgis.QgisAlgorithm import QgisAlgorithm
+
+
+class RasterSampling(QgisAlgorithm):
+
+    INPUT = 'INPUT'
+    RASTERCOPY = 'RASTERCOPY'
+    OUTPUT = 'OUTPUT'
+
+    def name(self):
+        return 'rastersampling'
+
+    def displayName(self):
+        return self.tr('Sample Raster Values')
+
+    def group(self):
+        return self.tr('Raster analysis')
+
+    def groupId(self):
+        return 'rasteranalysis'
+
+    def __init__(self):
+        super().__init__()
+
+    def initAlgorithm(self, config=None):
+        self.addParameter(
+            QgsProcessingParameterFeatureSource(
+                self.INPUT,
+                self.tr('Input Point Layer'),
+                [QgsProcessing.TypeVectorPoint]
+            )
+        )
+
+        self.addParameter(
+            QgsProcessingParameterMultipleLayers(
+                self.RASTERCOPY,
+                self.tr('Raster Layer to sample'),
+                QgsProcessing.TypeRaster
+            )
+        )
+
+        self.addParameter(
+            QgsProcessingParameterFeatureSink(
+                self.OUTPUT,
+                self.tr('Sampled Points')
+            )
+        )
+
+    def processAlgorithm(self, parameters, context, feedback):
+
+        source = self.parameterAsSource(
+            parameters,
+            self.INPUT,
+            context
+        )
+
+        sampled_rasters = self.parameterAsLayerList(
+            parameters,
+            self.RASTERCOPY,
+            context
+        )
+
+        if source is None:
+            raise QgsProcessingException(self.invalidSourceError(parameters, self.INPUT))
+
+        source_fields = source.fields()
+        raster_fields = QgsFields()
+
+        # append field to vector as rasterName_bandCount
+        for i in sampled_rasters:
+            for b in range(i.bandCount()):
+                raster_fields.append(QgsField(
+                    os.path.splitext(os.path.basename(i.source()))[0] + str('_{}'.format(b + 1)), QVariant.Double)
+                )
+
+        # combine all the vector fields
+        out_fields = QgsProcessingUtils.combineFields(source_fields, raster_fields)
+
+        (sink, dest_id) = self.parameterAsSink(
+            parameters,
+            self.OUTPUT,
+            context,
+            out_fields,
+            source.wkbType(),
+            source.sourceCrs()
+        )
+
+        if sink is None:
+            raise QgsProcessingException(self.invalidSinkError(parameters, self.OUTPUT))
+
+        total = 100.0 / source.featureCount() if source.featureCount() else 0
+        features = source.getFeatures()
+
+        for n, i in enumerate(source.getFeatures()):
+
+            for rr in sampled_rasters:
+
+                attrs = i.attributes()
+
+                if rr.bandCount() > 1:
+
+                    for b in range(rr.bandCount()):
+                        attrs.append(
+                            rr.dataProvider().identify(i.geometry().asPoint(),
```

Comment:
You'll need handling for geometryless features here, and for multipoint (with featureCount > 1) features.

For geometryless features you should set all the new attributes to NULL. For multipoint - I'm not sure. I'd probably just use something like `raise QgsProcessingException( self.tr('Feature {} is a multipoint features with {} parts. Run the 'multipart to single parts' algorithm on the layer first to split this feature into single point features.')`. Alternatively, the algorithm could do this automatically and export a single point feature per multipoint point input, with the raster values for that part. But I'd argue that approach could be dangerous, because if you don't know about this behaviour then your analysis might get screwed by having more features output then were input. That's why I'd go the explicit error message and requirement to use multipoint to singlepoint in advance.